src/acquisition/covidcast/direction_updater.py

```diff
@@ -144,17 +145,135 @@ def get_direction_impl(x, y):
+
+  # get the historical standard deviation of all signals and resolutions
+  rows = database.get_data_stdev_across_locations(
+    Constants.SIGNAL_STDEV_MAX_DAY)
+  data_stdevs = {}
+  for (source, signal, geo_type, aggregate_stdev) in rows:
+    if source not in data_stdevs:
+      data_stdevs[source] = {}
+    if signal not in data_stdevs[source]:
+      data_stdevs[source][signal] = {}
+    data_stdevs[source][signal][geo_type] = aggregate_stdev
+
+  # Dictionary to store the ids of rows with changed direction value
+  changed_rows = {-1.0: [], 0.0: [], 1.0: [], np.nan: []}
+
+  # looping over time-series
+  for ts_index, ts_key in enumerate(groupby_object.groups):
+    (
+        source,
+        signal,
+        geo_type,
+        geo_value,
+    ) = ts_key
+
+    ts_rows = groupby_object.get_group(ts_key).sort_values('time_value_datetime')
+    ts_rows['offsets'] = (ts_rows.time_value_datetime - ts_rows.time_value_datetime.iloc[0]).dt.days
+
+    # progress reporting for anyone debugging/watching the output
+    be_verbose = ts_index < 100
+    be_verbose = be_verbose or ts_index % 100 == 0
+    be_verbose = be_verbose or ts_index >= num_series - 100
+
+    if be_verbose:
+      msg = '[%d/%d] %s %s %s %s: span=%d--%d len=%d max_seconds_stale=%d'
+      args = (
+          ts_index + 1,
+          num_series,
+          source,
+          signal,
+          geo_type,
+          geo_value,
+          ts_rows.time_value.min(),
+          ts_rows.time_value.max(),
+          len(ts_rows),
+          ts_rows.timestamp1.max() - ts_rows.timestamp2.min()
+      )
+      print(msg % args)
+
+    offsets = ts_rows.offsets.values.astype(np.int64)
+    days = ts_rows.time_value.values.astype(np.int64)
+    values = ts_rows.value.values.astype(np.float64)
+    timestamp1s = ts_rows.timestamp1.values.astype(np.int64)
+    timestamp2s = ts_rows.timestamp2.values.astype(np.int64)
+
+    # create a direction classifier for this signal
+    data_stdev = data_stdevs[source][signal][geo_type]
+    slope_threshold = data_stdev * Constants.BASE_SLOPE_THRESHOLD
+
+    def get_direction_impl(x, y):
+      return direction_impl.get_direction(
+        x, y, n=Constants.SLOPE_STERR_SCALE, limit=slope_threshold)
+
+    # recompute any stale directions
+    days, directions = direction_impl.scan_timeseries(
+      offsets, days, values, timestamp1s, timestamp2s, get_direction_impl)
+
+    if be_verbose:
+      print(' computed %d direction updates' % len(directions))
+
+    # A DataFrame holding rows that potentially changed direction value
+    ts_pot_changed = ts_rows.set_index('time_value').loc[days]
+    ts_pot_changed['new_direction'] = np.array(directions, np.float64)
+
+    # is_eq_nan = ts_pot_changed.direction.isnull() & ts_pot_changed.new_direction.isnull()
+    # is_eq_num = ts_pot_changed.direction == ts_pot_changed.new_direction
+    # changed_mask = ~(is_eq_nan | is_eq_num)
+    # ts_changed = ts_pot_changed[changed_mask]
+
```

Comment:
Should we remove these if not needed anymore?

Reply:
I agree, I considered updating only the rows where the value actually changed, it turned out that this increases overall runtime, so I didn't end up using it.